From: Marek Skrobacki <marek.skrobacki@rackspace.co.uk>
Subject: [PATCH] add support for dhcp-allowed-srvids

Allow a DHCPREQUEST's Server ID (option 54) to be accepted even when it
doesn't match an address on the local interface, for cases where dnsmasq
runs behind a load balancer or inside a container, provided the value is
explicitly allow-listed via --dhcp-allowed-srvids.

Rebased against dnsmasq 2.93 (as packaged in Debian forky).
---
diff --git a/man/dnsmasq.8 b/man/dnsmasq.8
index 5b01e63..099c65f 100644
--- a/man/dnsmasq.8
+++ b/man/dnsmasq.8
@@ -2094,6 +2094,25 @@ form is used, there must be a route to all of the addresses configured on the in
 The two-address form of shared-network is also usable with a DHCP relay: the first address
 is the address of the relay and the second, as before, specifies an extra subnet which
 addresses may be allocated from.
+.TP
+.B --dhcp-allowed-srvids[=<ip addr>]
+(IPv4 Only) If the DHCP server is running inside a container or behind a load balancer, the
+DHCPREQUEST arriving at dnsmasq for processing may have a Server ID (option 54)
+configured with an IP address that is not assigned to the local interface. In
+this case, dnsmasq will check if the 'Server Identifier Override' option was
+set in the incoming packet.
+.IP
+If it was not set, the packet is dropped.
+.IP
+If it was set, dnsmasq evaluates the Server ID against the value provided in \'Server ID Override\' suboption 
+11, as outlined in RFC5107.
+.IP
+In both cases, there is no match against the \'backend\' IP address configured on the interface.
+.IP
+The \fB\-\-dhcp\-allowed\-srvids\fP option allows disabling this security
+mechanism for specific address(es). When enabled, the incoming DHCPREQUEST is
+evaluated against the provided value(s) instead of the addresses configured on
+the local interfaces.
 
 .TP
 .B \-s, --domain=<domain>[[,<address range>[,local]]|<interface>]
diff --git a/src/dnsmasq.h b/src/dnsmasq.h
index ebb8d15..ec9b0c4 100644
--- a/src/dnsmasq.h
+++ b/src/dnsmasq.h
@@ -1236,11 +1236,13 @@ extern struct daemon {
   struct pxe_service *pxe_services;
   struct tag_if *tag_if; 
   struct addr_list *override_relays;
+  struct addr_list *allowed_srvids;
   struct dhcp_relay *relay4, *relay6;
   struct delay_config *delay_conf;
   int override;
   int enable_pxe;
   int doing_ra, doing_dhcp6;
+  int allowing_custom_srvids;
   struct dhcp_netid_list *dhcp_ignore, *dhcp_ignore_names, *dhcp_gen_names; 
   struct dhcp_netid_list *force_broadcast, *bootp_dynamic;
   struct hostsfile *dhcp_hosts_file, *dhcp_opts_file;
diff --git a/src/option.c b/src/option.c
index 274732f..75c954a 100644
--- a/src/option.c
+++ b/src/option.c
@@ -201,6 +201,7 @@ struct myoption {
 #define LOPT_LEASEQUERY    389
 #define LOPT_SPLIT_RELAY   390
 #define LOPT_LOG_MALLOC    391
+#define LOPT_DHCP_AL_SVID  392
 
 #ifdef HAVE_GETOPT_LONG
 static const struct option opts[] =  
@@ -403,6 +404,7 @@ static const struct myoption opts[] =
     { "max-tcp-connections", 1, 0, LOPT_MAX_PROCS },
     { "leasequery", 2, 0, LOPT_LEASEQUERY },
     { "log-malloc", 0, 0, LOPT_LOG_MALLOC },
+    { "dhcp-allowed-srvids", 1, 0, LOPT_DHCP_AL_SVID },
     { NULL, 0, 0, 0 }
   };
 
@@ -612,8 +614,9 @@ static struct {
   { LOPT_CACHE_RR, ARG_DUP, "<RR-type>", gettext_noop("Cache this DNS resource record type."), NULL },
   { LOPT_MAX_PROCS, ARG_ONE, "<integer>", gettext_noop("Maximum number of concurrent tcp connections."), NULL },
   { LOPT_LOG_MALLOC, OPT_LOG_MALLOC, NULL, gettext_noop("Log memory allocation for debugging."), NULL },
+  { LOPT_DHCP_AL_SVID, ARG_DUP, "[=<ipaddr>]...", gettext_noop("Allow these ServerIDs"), NULL },
   { 0, 0, NULL, NULL, NULL }
-}; 
+};
 
 /* We hide metacharacters in quoted strings by mapping them into the ASCII control
    character space. Note that the \0, \t \b \r \033 and \n characters are carefully placed in the
@@ -4809,6 +4812,18 @@ static int one_opt(int option, char *arg, char *errstr, char *gen_err, int comma
 	
 	break;
       }
+    case LOPT_DHCP_AL_SVID:
+	daemon->allowing_custom_srvids = 1;
+	while (arg) {
+	struct addr_list *new = opt_malloc(sizeof(struct addr_list));
+	comma = split(arg);
+	if (!(inet_pton(AF_INET, arg, &new->addr) > 0))
+	  ret_err_free(_("bad dhcp-allowed-srvids address"), new);
+	new->next = daemon->allowed_srvids;
+	daemon->allowed_srvids = new;
+	arg = comma;
+	}
+	break;
 
 #endif
       
diff --git a/src/rfc2131.c b/src/rfc2131.c
index 7b58e69..98f3ff4 100644
--- a/src/rfc2131.c
+++ b/src/rfc2131.c
@@ -1407,8 +1407,22 @@ size_t dhcp_reply(struct dhcp_context *context, char *iface_name, int int_index,
 	      
 	      if (override.s_addr != 0)
 		{
-		  if (option_addr(opt).s_addr != override.s_addr)
-		    return 0;
+		  if (option_addr(opt).s_addr != override.s_addr) {
+		    if (daemon->allowing_custom_srvids) {
+			struct addr_list *l;
+			for (l = daemon->allowed_srvids; l; l = l->next)
+			  if (l->addr.s_addr == option_addr(opt).s_addr) {
+			    inet_ntop(AF_INET, &l->addr.s_addr, daemon->addrbuff, ADDRSTRLEN);
+			    my_syslog(MS_DHCP | LOG_DEBUG, _("ServerID %s is explicitly allowed."),
+				        daemon->addrbuff);
+			    break;
+			  }
+			if (!l)
+			  return 0;
+		    }
+		   else
+			return 0;
+		  }
 		}
 	      else 
 		{
@@ -1435,12 +1449,28 @@ size_t dhcp_reply(struct dhcp_context *context, char *iface_name, int int_index,
 			override = intr->addr.in.sin_addr;
 		      else
 			{
-			  /* In auth mode, a REQUEST sent to the wrong server
-			     should be faulted, so that the client establishes 
-			     communication with us, otherwise, silently ignore. */
-			  if (!option_bool(OPT_AUTHORITATIVE))
-			    return 0;
-			  message = _("wrong server-ID");
+			  if (daemon->allowing_custom_srvids) {
+				my_syslog(MS_DHCP | LOG_DEBUG, _("checking allowed custom serverids"));
+				struct addr_list *l;
+				for (l = daemon->allowed_srvids; l; l = l->next)
+				  if (l->addr.s_addr == option_addr(opt).s_addr)
+				    break;
+				if (l) {
+		              inet_ntop(AF_INET, &l->addr.s_addr, daemon->addrbuff, ADDRSTRLEN);
+				  my_syslog(MS_DHCP | LOG_DEBUG, _("ServerID %s is explicitly allowed."),
+						daemon->addrbuff);
+				  override = option_addr(opt);
+				}
+				else
+				{
+				  /* In auth mode, a REQUEST sent to the wrong server
+				  should be faulted, so that the client establishes
+				  communication with us, otherwise, silently ignore. */
+				  if (!option_bool(OPT_AUTHORITATIVE))
+				    return 0;
+				  message = _("wrong server-ID");
+				}
+			  }
 			}
 		    }
 		}
